### PR TITLE
Fixes #16303: evar defined before its instance was checked, leading to error message trying to unexpectingly display an already defined evar

### LINKED ISSUE #16303: Uncaught exception Not_found Error
#### Description of the problem
I just accidentally discovered this issue (see the code), and it asked me to report.
```

Fact uip_nat {n : nat} (e : n = n) : e = eq_refl.
  Proof. 
   refine match e with 
    | eq_refl => eq_refl
    end. 

<in exception printer>:<original exception:Anomaly
                                           "Uncaught exception Not_found."
Please report at http://coq.inria.fr/bugs/.
```

#### Coq Version
```
The Coq Proof Assistant, version 8.15.2
compiled with OCaml 4.12.1
```

### PR DESCRIPTION
**Pull request by herbelin**
This improves commit 26a5b65df from PR #14797.

Fixes #16303

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

### CODE REVIEW

File: pretyping/evarsolve.ml [hunk 1]
```diff
@@ -1486,8 +1486,8 @@ let solve_candidates unify flags env evd (evk,argsv) rhs =
           (* solve_candidates might have been called recursively in the mean *)
           (* time and the evar been solved by the filtering process *)
          if Evd.is_undefined evd evk then
-           let evd' = Evd.define evk c evd in
-             check_evar_instance unify flags env evd' evk c
+           let evd' = check_evar_instance unify flags env evd evk c in
+           Evd.define evk c evd'
```
Comment (gares) [inline pretyping/evarsolve.ml:1490]: Shouldn't we have a single function which both check and assigns the evar?
I did some git grep and it seems doable, all uses of `check_evar_instance` are next to an `Evd.define`

Comment (herbelin) [reply pretyping/evarsolve.ml:1490]: > Shouldn't we have a single function which both check and assigns the evar? I did some git grep and it seems doable, all uses of `check_evar_instance` are next to an `Evd.define`

Indeed, that would make a lot of sense.

**Comment by herbelin** [comment]
@coqbot ci minimize

**Comment by herbelin** [comment]
@coqbot ci minimize

**Comment by herbelin** [comment]
@coqbot ci minimize

**Comment by herbelin** [comment]
@coqbot ci minimize

**Comment by herbelin** [comment]
The HoTT failure is an endless loop and the minimizer, unsatisfied with the observed timeout, restarts a new minimization at regular new times. It could be stopped.

**Comment by JasonGross** [comment]
It seems like coqtop hangs rather than stack overflowing, and so the minimizer never makes it past the initial step because I don't impose a timeout on splitting.

**Comment by herbelin** [comment]
Alternative fix in #16869.

Eventually, the PR could add its tests to the test-suite though.

**Pull request closed without merging**